### PR TITLE
[FLINK-3463] implement calc translation

### PR DESCRIPTION
**Pull request by vasia**
This PR implements the `DataSetCalcRule`, so that filters, projections, and their combinations are now translated to `FlinkCalc`s. `FlinkFilterRule`, `FlinkProjectRule`, `DataSetFilterRule`, and `DataSetProjectRule` are removed.
@twalthr @fhueske let me know if that's what you had in mind :)

### CODE REVIEW

**Comment by fhueske** [comment]
I think we can also remove `FlinkProject` and `FlinkFilter` since there are no rules left that translate to these RelNodes.
Otherwise looks good to me.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/table/test/CalcITCase.scala [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import org.apache.flink.test.util.MultipleProgramsTestBase
+import org.apache.flink.test.util.MultipleProgramsTestBase.TestExecutionMode
+import org.apache.flink.test.util.TestBaseUtils
+import org.junit._
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+import scala.collection.JavaConverters._
+
+@RunWith(classOf[Parameterized])
+class CalcITCase(mode: TestExecutionMode) extends MultipleProgramsTestBase(mode) {
```
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/table/test/CalcITCase.scala]: Could you change `MultipleProgramsTestBase` to `TableProgramsTestBase`? Then we can also test different configurations.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/table/test/CalcITCase.scala [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+  @Test
+  def testSimpleCalc(): Unit = {
+
+    val env = ExecutionEnvironment.getExecutionEnvironment
+    val t = CollectionDataSets.get3TupleDataSet(env).toTable
```
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/table/test/CalcITCase.scala]: `toTable` can then get the config through `toTable(getConfig)`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataset/DataSetCalcRule.scala [hunk 3]
```diff
@@ -37,13 +45,90 @@ class DataSetCalcRule
...
+          s"""
+            |${projection.code}
+            |${generator.collectorTerm}.collect(${projection.resultTerm});
+            |""".stripMargin
+        }
+        else {
+          val filterCondition = generator.generateExpression(
+              calcProgram.expandLocalRef(calcProgram.getCondition))
+          if (projection == null) {
+          // only filter
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataset/DataSetCalcRule.scala]: Just a small comment: Maybe I'm alone in this opinion, but I think it's nicer to have the comment above the `if`, it improves the reading flow (for me at least). Then you also don't have wrong indention, like in this line.

**Comment by twalthr** [comment]
@vasia PR looks good. Cool, that you found a way to trigger a calc without custom rules.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/table/test/CalcITCase.scala (hunk 2 above)
Comment (vasia) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/table/test/CalcITCase.scala]: Hey @twalthr, `toTable(getConfig)` doesn't seem to exist. Could you take a look in the updated test and let me know if it's OK or there's something I'm missing? Thanks!

**Comment by vasia** [comment]
Thanks @fhueske  and @twalthr for the comments! Apart from a small doubt with the test, I've addressed the rest.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/table/test/CalcITCase.scala (hunk 2 above)
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/table/test/CalcITCase.scala]: @vasia Sorry, I mixed up `toTable` and `toDataSet`. `toDataSet[Row](config)` exists.

Comment (vasia) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/table/test/CalcITCase.scala]: Thanks!

**Comment by vasia** [comment]
I've updated the PR.
Changing the test to `TableProgramsTestBase ` actually made the `TestCalcWithAggregation` test fail for `config=EFFICIENT` on a `ClassCastExcpetion` (`Tuple3` instead of `Row`). I've fixed that by forwarding the expectedType in `DataSetFlatMap`. Let me know if that's OK. Thanks!

**Comment by twalthr** [comment]
Forwarding the expected type is not an option, because you don't know what the previous operator produces. If the previous operator produces `Tuple5` and you force it to use `Tuple3` it won't work. I will look into this issue and write some documentation about the interfaces so that it becomes clearer.

**Comment by twalthr** [comment]
I have opened a PR (#1709) that adds some documentation and does some type checking for aggregates. The problem is that aggregates do not support other types than `Row`s. They therefore expect the `Row` type, which was not passed to input operators.
You can remove the changes in `DataSetFlatMap`. #1709 should solve you issue.
Btw, `TestCalcWithAggregation` should be renamed to `testCalcWithAggregation` ;-)

**Comment by vasia** [comment]
Thanks @twalthr! You PR solves the problem indeed :)
If no objections, I will merge both #1709 and this one rebased on your changes.

**Comment by twalthr** [comment]
+1 for merging

**Comment by vasia** [comment]
Thanks! I'll merge both once Travis passes.

**Pull request closed without merging**